src/Xamarin.Android.Build.Tasks/Utilities/TypeMappingNativeAssemblyGenerator.cs

```diff
@@ -1,75 +1,240 @@
 
-		public TypeMappingNativeAssemblyGenerator (NativeAssemblerTargetProvider targetProvider, NativeAssemblyDataStream dataStream, string dataFileName, uint dataSize, string mappingFieldName)
-			: base (targetProvider)
+		public TypeMappingNativeAssemblyGenerator (NativeAssemblerTargetProvider targetProvider, NativeTypeMappingData mappingData, string baseFileName, bool sharedBitsWritten, bool sharedIncludeUsesAbiPrefix = false)
+			: base (targetProvider, baseFileName, sharedIncludeUsesAbiPrefix)
 		{
-			if (dataStream == null)
-				throw new ArgumentNullException (nameof (dataStream));
-			if (String.IsNullOrEmpty (dataFileName))
-				throw new ArgumentException ("must not be null or empty", nameof (dataFileName));
-			if (String.IsNullOrEmpty (mappingFieldName))
-				throw new ArgumentException ("must not be null or empty", nameof (mappingFieldName));
-
-			this.dataStream = dataStream;
-			this.dataFileName = dataFileName;
-			this.dataSize = dataSize;
-			this.mappingFieldName = mappingFieldName;
+			this.mappingData = mappingData ?? throw new ArgumentNullException (nameof (mappingData));
+
+			if (String.IsNullOrEmpty (baseFileName))
+				throw new ArgumentException("must not be null or empty", nameof (baseFileName));
+
+			this.baseFileName = baseFileName;
+			this.sharedBitsWritten = sharedIncludeUsesAbiPrefix ? false : sharedBitsWritten;
 		}
 
-		public bool EmbedAssemblies { get; set; }
+		protected override void WriteSymbols (StreamWriter output)
+		{
+			output.WriteLine ();
+			WriteHeaderField (output, "map_module_count", mappingData.MapModuleCount);
+
+			output.WriteLine ();
+			WriteHeaderField (output, "java_type_count", mappingData.JavaTypeCount);
+
+			output.WriteLine ();
+			WriteHeaderField (output, "java_name_width", mappingData.JavaNameWidth);
+
+			bool haveAssemblyNames = mappingData.AssemblyNames.Count > 0;
+			bool haveModules = mappingData.Modules.Count > 0;
 
-		protected override void WriteFileHeader (StreamWriter output, string outputFileName)
+			output.WriteLine ();
+			if (haveAssemblyNames) {
+				output.WriteLine ($"{Indent}.include{Indent}\"{Path.GetFileName (SharedIncludeFile)}\"");
+			} else {
+				WriteCommentLine (output, $"No shared data present, {Path.GetFileName (SharedIncludeFile)} not generated");
+			}
+
+			if (haveModules) {
+				output.WriteLine ($"{Indent}.include{Indent}\"{Path.GetFileName (TypemapsIncludeFile)}\"");
+			} else {
+				WriteCommentLine (output, $"No modules defined, {Path.GetFileName (TypemapsIncludeFile)} not generated");
+			}
+			output.WriteLine ();
+
+			if (!sharedBitsWritten && haveAssemblyNames) {
+				using (var ms = new MemoryStream ()) {
+					using (var sharedOutput = new StreamWriter (ms, output.Encoding)) {
+						WriteAssemblyNames (sharedOutput);
+						sharedOutput.Flush ();
+						MonoAndroidHelper.CopyIfStreamChanged (ms, SharedIncludeFile);
+					}
+				}
+			}
+
+			if (haveModules) {
+				using (var ms = new MemoryStream ()) {
```

Comment:
Hmm, maybe not easy to reuse the `MemoryStream` here without making the code ugly.

I've thought about making an `ObjectPool<T>` that could be used for `MemoryStream`, something like here: https://docs.microsoft.com/en-us/dotnet/standard/collections/thread-safe/how-to-create-an-object-pool

Maybe it would be worth it if used across the entire build? The underlying arrays are probably on the LOH.

This PR is probably fine as-is.